keps/sig-storage/20200120-generic-data-populators.md

```diff
@@ -0,0 +1,309 @@
+repositories dedicated to that purpose, often with various mechanisms for retrieving
+them efficiently that are external to Kubernetes.
+
+One way to achieve this is to represent disk images as custom resources that point to
+the image repository, and to allow creation of PVCs from these custom resources such
+that the volumes come pre-populated with the correct data. Efficient population of the
+data could be left up to a purpose-built controller that knows how to get the bits
+where they need to be with minimal I/O.
+
+#### Backup/Restore
+
+Without getting into the details of how backup/restore should be implemented, it's
+clear that whatever design one chooses, a necessary step is to have the user
+(or higher level controller) create a PVC that points to the backup they want to
+restore, and have the data appear in that volume somehow.
+
+One can imagine backups simply being a special case of snapshots, in which case the
+existing design is sufficient, but if you want anything more sophisticated, there
+will inevitably be a new object type that represents a backup. While it's arguable
+that backup should be something CSI plugins should be exclusively responsible for,
+one can also argue that generic backup tools should also exist which can backup
+and restore all kind of volumes. Those tools will be apart from CSI plugins and
+yet need a way to populate volumes with restored volumes.
+
+It's also likely that multiple backup/restore implementations will be developed,
+and it's not a good idea to pick a winner at the Kubernetes API layer. It makes
+more sense to enable developers to try different approaches by making the API allow
+restoring from various kinds of things. 
+
+### Implementation Details/Notes/Constraints
+
+As noted above, the proposal is extremely simple -- just remove the validation on
+the `DataSource` field. This raises the question of WHAT will happen when users
+put new things in that field, and HOW populators will actually work with so small
+a change.
+
+It's first important to note that only consumers of the `DataSource` field today
+are the various dynamic provisioners, most notably the external-provisioner CSI
+sidecar. If the external-provisioner sidecar sees a data source it doesn't
+understand, it simply ignores the request, which is both important for forward
+compatibility, and also perfect for the purposes of a data populator. This allows
+developers to add new types of data sources that the dynamic provisioners will
+simply ignore, enabling a different controller to see these objects and respond
+to them.
+
+The way I expect most data populators to work in practice is that they will see
+a request for a PVC `foo` with a data source of `bar`. The populator will create
+a shadow PVC `foo'` in another namespace with all of the same spec except the
+data source will be empty. The dynamic provisioner will ignore `foo` but will
+respond to `foo'` by creating an empty PV. The populator can create a pod
```

Comment:
Without taints, how do you prevent PVC and the empty PV from being bound to each other and being used by a pod?

Reply:
If it's not clear, there are two PVCs -- foo and foo-prime. The PV initially binds to foo-prime. After the PV is populated, we delete foo-prime and rebind the PV to foo.

Reply:
Would you be able to include an example workflow/manifests, you've called out the multiple PVCs but I'm not sure how you're intending that to work from a users perspective, or from the ns transfer perspective.

Reply:
The example I provided is in the hello-populator implementation. I could try to replicate that example in slides to better illustrate how it works, but KEPs don't include slides...

Reply:
Please - excruciating detail is needed

Reply:
I can make a slide deck on this subject... Let me start here with a description of what the hello-populator example code does.

The hello-populator creates a second PV with all of the same details as the first PV, but bound to the other PVC. This takes advantage to the fact that two different PV objects can point to exactly the same storage. Because the original PV is bound to a PVC that the populator created -- we can ensure that there are definitely no pods attached to the PVC at the time this happens, avoiding any dangers related to multiple attachments when that shouldn't be allowed. And of course we change the retentionPolicy on the original PV to Retain before we delete it so nothing happens to the actual storage while we clean up the temporary k8s objects.

There are other flavors of this -- you can delete the PV first and then re-create it bound to the other PVC. This make recovering from errors a little harder because you have to persist the state needed to clone the PV object somewhere.
You can delete the shadow PVC before cloning its PV, but leave a finalizer on it to keep it around, to make extra sure no new pods can attach to it while there are two PVs pointing to the same storage. I didn't implement this extra safety-step, but it's conceptually not hard. I also did some experiments with simply modifying the PV object directly to force it to rebind. If you set its rentionPolicy to Retain and delete the PVC, you can get the bind controller to think it's okay to bind it again. I forget why, but this method did cause some problems which is why I used the two PV approach for my sample.

Reply:
Actually I just re-created my old experiment where I directly modify the PV to rebind it, and nothing bad happened, so that approach may be preferable.